### ISSUE TITLE
Don't download duplicate nightly

### ISSUE THREAD

**Issue by durka** [rust-lang/rustup#1366]
`rustup update nightly-YYYY-MM-DD` shouldn't need to update anything if the currently installed `nightly` is already YYYY-MM-DD. This also occurs when using a `rust-toolchain` file.

**Comment by Boscop** [comment]
Yes, please don't download it again, some of us have very slow internet sometimes..

>   3.4 MiB /  51.5 MiB (  7 %)  16.0 KiB/s ETA:  51 min 17 s

**Comment by zoumi** [comment]
Can we install rust without rust-docs? It was useless for me, but take so much time to download & install it.

**Comment by das-sein** [comment]
This might take some time to resolve. As far as I can tell, it would require including the date in the toolchain names and storing them with those included dates. https://github.com/rust-lang/rustup.rs/blob/master/src/rustup-dist/src/dist.rs#L254 Not sure if this is desirable behavior as this would alter something like `rustup default nightly` to something like `rustup default nightly-2019-02-05`. Hopefully one of the primary contributors will respond and correct me.

**Comment by junderw** [comment]
This is also an issue for stable (or any toolchain, really)

I have stable installed with is currently 1.70.0.....

I built a project with rust-toolchain of `1.70` which downloads a whole new 1.2 GB for a duplicate toolchain.

I then built another project which has a rust-toolchain of `1.70.0` which downloads a whole new 1.2 GB for a duplicate toolchain.

I think the toolchain resolver needs to do some searching of locally installed toolchains first.

ie.

1. `stable` will lookup the latest stable, It will find 1.70.0, then search locally for any 1.70, 1.70.0 toolchains, and symlink to that if possible, download if not.
2. `1.70` will lookup the latest 1.70.x, find 1.70.0 and then search for it in the `stable` and any `1.70.0` toolchains first, and create a symlink if already available, download if not.
3. `1.70.0` will look at the local `1.70` and `stable` to see if 1.70.0 exists, and create symlink if available, download if not.
4. When updating `stable` or `1.70` and it bumps the version, the actual non-symlink data folder should be renamed to 1.70.0 toolchain and left there.
5. So whichever was installed first is the actual data, then the other two just symlink to it, and if either of the symlinks updates to a higher version, it bumps the data folder down to the level below that and updates the symlink to the other while updating (and downloading) self.

Example:

1. Install `1.70` and it gets "1.70.0" from the server, looks locally, finds no stable or 1.70.x toolchains, downloads it.
2. Install `1.70.0` and it recognizes itself within the `1.70` toolchain and creates a symlink.
3. Install `stable` and it gets "1.70.0" response from the server, then looks locally and finds it in `1.70` and creates a symlink.
4. Update `1.70` and it finds "1.70.1", looks locally, doesn't find it, finds `1.70.0`, notices it's a symlink, deletes symlink and moves self to 1.70.0, finds stable is also pointing to self, points it to 1.70.0 as well. Then downloads 1.70.1 and uses it.
5. Update `stable` and it finds "1.70.1", looks locall, finds it in `1.70` and updates its symlink.

So the flow is:

1. Is specific version? If no, search online to find latest specific version that matches toolchain.
2. Take specific version, and search local toolchains from specific to general (`1.70.0` to `stable`) to see if the specific version is already locally installed.
3. Once found (either locally, or "needs download"), if this is an update (ie. you will be overwriting an existing self. Skip to 5 if not an update) check to see if any of the other toolchains (specific to general) are pointing to self. Take the first one you find that is pointing to self and delete their symlink and rename self to their toolchain.
4. If none were pointing to self, delete/overwrite old self is ok.
5. Delete self and either create symlink to the found new self or download new self.

**Comment by rbtcollins** [comment]
There are some design points to manage in here.

Firstly, 1.70 and 1.70.0 are not the same toolchain. They might be on a given date, but once 1.70.1 releases, they are not.

Secondly, the installed components are currently defined as separate for each toolchain; adding a component that isn't expected might be strange and conceivable a bug for some developers - and turning minor versions into symlinks to concrete versions will have that effect, as you would no longer have separate component lists. 

It will also be confusing to IDE tooling that may be adding components or updating toolchains, if they can run into our existing concurrency bugs *when making changes to separate toolchains* - so an API change in this area will leak out to all automation wrapped around rustup.

I don't know what the right answer looks like, but I think it needs to reason through all those points.

**Comment by junderw** [comment]
> Firstly

I agree. This is the reason why the slightly confusing symlink dance is necessary, and (assuming it works as designed) should account for this point barring concurrency issues.

> Secondly

I am having trouble groking any situations where we would be adding a component that isn't already possible by "using another project with the same toolchain but needing more tools".

If we have `1.70` installed first, used for Project A, then for Project B we "install" (via symlink) `1.70.0` afterward (and at the time it is the latest), the only "unexpected" thing would be "I thought I was going to install it fresh but it turns out I have all these tools and targets already installed."

However, if Project A used `1.70.0` and requires all those tools it's the same state which Project B will find.

Then maybe Project B requires to add a new target, but I'm not sure how that would be detrimental to Project A.

I would love to hear more about this perceived issue, as I am not too familiar with rustup's various issues.

> confusing to IDE tooling ... run into our existing concurrency bugs when making changes to separate toolchains

This seems like the biggest issue, and since I am not very familiar with rustup, I don't have much to say on this. It seems like the only stopper I can semi-grok.

---

That said, most all of these issues could be "solved" by making the feature experimental / opt-in so that people with slow internet can benefit from it (with the added caution necessitated by the potential issues).

Then it can be moved from experimental to stable / default-on once the major concurrency bugs are fixed.

Thanks for your insight!
